src/emit/polkadot/mod.rs

```diff
@@ -200,12 +200,13 @@ impl PolkadotTarget {
             .builder
             .build_call(binary.module.get_function("__init_heap").unwrap(), &[], "");
 
-        let scratch_buf = binary.scratch.unwrap().as_pointer_value();
-        let scratch_len = binary.scratch_len.unwrap().as_pointer_value();
+        let input_buf_ty = binary.context.i8_type().array_type(SCRATCH_SIZE);
```

Comment:
I don't think this is a good idea. You're allocating 32KiB on the stack, and this is never freed. So, for the majority of the execution this memory is just wasted.

Also, after your change, whenever `msg.data`  is accessed, the input buffer is copied again, even though there already is a copy on the stack.

You could use another static buffer (just like the scratch buffer) for the input, and avoid the allocation cost and it can re-used whenever `msg.data` or `msg.sig` is used.

Reply:
Alternatively, `msg.sender` could use the stack rather than the scratch buffer

Reply:
> I don't think this is a good idea. You're allocating 32KiB on the stack, and this is never freed. So, for the majority of the execution this memory is just wasted.

How is using the scratch buffer different?

> Also, after your change, whenever msg.data is accessed, the input buffer is copied again, even though there already is a copy on the stack.
>
> You could use another static buffer (just like the scratch buffer) for the input, and avoid the allocation cost and it can re-used whenever msg.data or msg.sig is used.

I think a dedicated buffer for the input is a good idea. However I'm not concerned about the cost of the stack allocation. I didn't check but I'd be very surprised if its expensive (even more if it was so expensive that it's a concern; we use it for API calls which are super costly anyways). Ideally we can just :axe: the scratch buffer completely, it's causing multiple problems like this.

Reply:
Or, as discussed, we can move the call to the storage initializer before the input reading. This will avoid the problem too.

Reply:
@seanyoung you were right, I solved it differently. WDYT?